contracts/RMRK/utils/RMRKBulkWriter.sol

```diff
@@ -0,0 +1,136 @@
+     * @param collection Address of the collection that this contract is managing.
+     * @param tokenId ID of the token we are managing.
+     */
+    modifier onlyTokenOwner(address collection, uint256 tokenId) {
+        _checkTokenOwner(collection, tokenId);
+        _;
+    }
+
+    /**
+     * @notice Initializes the contract
+     */
+    constructor() {}
+
+    /**
+     * @notice Replaces the current equipped child on the asset and slot combination with the given one
+     * @dev The `IntakeEquip` stuct contains the following data:
+     *  [
+     *      tokenId,
+     *      childIndex,
+     *      assetId,
+     *      slotPartId,
+     *      childAssetId
+     *  ]
+     * @param collection Address of the collection that this contract is managing.
+     * @param data An `IntakeEquip` struct specifying the equip data.
+     */
+    function replaceEquip(
+        address collection,
+        IERC6220.IntakeEquip memory data
+    ) public onlyTokenOwner(collection, data.tokenId) {
+        IERC6220(collection).unequip(
+            data.tokenId,
+            data.assetId,
+            data.slotPartId
+        );
+        IERC6220(collection).equip(data);
+    }
+
+    /**
+     * @notice Performs multiple unequip and equip operations
+     * @dev Unequip operations must run first
+     * @dev tokenId is included as a parameter to be able to do a single check for ownership
+     * @dev Every tokenId in the `IntakeEquip` structs must match the tokenId passed in
+     * @dev The `IntakeUnequip` stuct contains the following data:
+     *  [
+     *      assetId,
+     *      slotPartId,
+     *  ]
+     * @dev The `IntakeEquip` stuct contains the following data:
+     *  [
+     *      tokenId,
+     *      childIndex,
+     *      assetId,
+     *      slotPartId,
+     *      childAssetId
+     *  ]
+     * @param collection Address of the collection that this contract is managing.
+     * @param tokenId ID of the token we are managing.
+     * @param unequips[] An array of `IntakeUnequip` structs specifying the slots to unequip.
+     * @param equips[] An array of `IntakeEquip` structs specifying the slots to equip.
+     */
+    function bulkEquip(
+        address collection,
+        uint256 tokenId,
+        IntakeUnequip[] memory unequips,
+        IERC6220.IntakeEquip[] memory equips
+    ) public onlyTokenOwner(collection, tokenId) {
+        uint256 length = unequips.length;
+        for (uint256 i = 0; i < length; ) {
+            IERC6220(collection).unequip(
+                tokenId,
+                unequips[i].assetId,
+                unequips[i].slotPartId
+            );
+            unchecked {
+                ++i;
+            }
+        }
+        length = equips.length;
+        for (uint256 i = 0; i < length; ) {
+            if (equips[i].tokenId != tokenId) {
+                revert RMRKCanOnlyDoBulkOperationsWithOneTokenAtATime();
+            }
```

Comment:
I'd add the same check to the unequip actions 🤔

Reply:
Unequips does not have token Id, I take it directly from the param. So nothing to compare there.

Reply:
What about the length of `unequips`?

Reply:
What about it?

Reply:
Ahhh, brain fart moment 😂